File: src/main/java/com/lessmarkup/engine/data/QueryBuilderImpl.java
package com.lessmarkup.engine.data;

import com.lessmarkup.Constants;
import com.lessmarkup.engine.data.dialects.DatabaseLanguageDialect;
import com.lessmarkup.engine.data.dialects.DatabaseLanguageDialectFactory;
import com.lessmarkup.framework.helpers.DependencyResolver;
import com.lessmarkup.framework.helpers.PropertyDescriptor;
import com.lessmarkup.framework.helpers.StringHelper;
import com.lessmarkup.framework.helpers.TypeHelper;
import com.lessmarkup.interfaces.data.DataObject;
import com.lessmarkup.interfaces.data.OptionalBoolean;
import com.lessmarkup.interfaces.data.QueryBuilder;
import com.lessmarkup.interfaces.exceptions.CommonException;
import com.lessmarkup.interfaces.exceptions.DatabaseException;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.ResultSetMetaData;
import java.sql.SQLException;
import java.sql.Timestamp;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.*;

class QueryBuilderImpl implements QueryBuilder {
    
    private Connection connection;
    private DatabaseLanguageDialect dialect;
    
    private final StringBuilder commandText = new StringBuilder();
    private String select = "*";
    private final StringBuilder where = new StringBuilder();
    private final StringBuilder orderBy = new StringBuilder();
    private String limit = "";
    private final List<Object> parameters = new ArrayList<>();
    
    void setConnection(Connection connection) {
        this.connection = connection;
        try {
            this.dialect = DatabaseLanguageDialectFactory.createDialect(this.connection);
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
    }

    @Override
    public <T extends DataObject> QueryBuilder from(Class<T> type, String name) {
        TableMetadata metadata = DomainModelImpl.getMetadata(type);
        if (metadata == null) {
            throw new IllegalArgumentException();
        }
        this.commandText.append(String.format(" FROM %s %s", this.dialect.decorateName(metadata.getName()), name != null ? name : ""));
        return this;
    }
    
    private <T extends DataObject> QueryBuilder anyJoin(Class<T> type, String joinType, String name, String on) {
        TableMetadata metadata = DomainModelImpl.getMetadata(type);
        if (metadata == null) {
            throw new IllegalArgumentException();
        }
        this.commandText.append(String.format(" %sJOIN %s %s ON %s", joinType, this.dialect.decorateName(metadata.getName()), name, on));
        return this;
    }
    
    private String processStringWithParameters(String sql, Object[] args) {
        if (args.length == 0) {
            return sql;
        }
        
        int pos = 0;
        int argIndex = 0;
        
        StringBuilder ret = new StringBuilder();
        
        for (;;) {
            int newPos = sql.indexOf('$', pos);
            
            if (newPos < 0) {
                ret.append(sql.substring(pos));
                break;
            }
            
            if (newPos > pos) {
                ret.append(sql.substring(pos, newPos));
            }
            
            pos = newPos;
            
            boolean isTableName = pos + 1 < sql.length() && sql.charAt(pos+1) == '-';
            
            if (isTableName) {
                int pos1 = pos+2;
                for (;pos1 < sql.length() && Character.isAlphabetic(sql.charAt(pos1)); pos1++) {
                }
                String tableName = sql.substring(pos+2, pos1);
                tableName = DomainModelImpl.getMetadata(tableName).getName();
                ret.append(tableName);
                pos = pos1;
                continue;
            }
            
            if (pos + 1 < sql.length() && Character.isDigit(sql.charAt(pos+1))) {
                int start = pos+1;
                pos += 2;
                for (; pos < sql.length() && Character.isDigit(sql.charAt(pos)); pos++) {
                }
                int parameterIndex = Integer.decode(sql.substring(start, pos));
                ret.append('?');
                this.parameters.add(args[parameterIndex]);
                continue;
            }
            
            ret.append('?');
            this.parameters.add(args[argIndex++]);
            pos++;
        }
        
        return ret.toString();
    }

    @Override
    public <T extends DataObject> QueryBuilder join(Class<T> type, String name, String on) {
        return anyJoin(type, "", name, on);
    }

    @Override
    public <T extends DataObject> QueryBuilder leftJoin(Class<T> type, String name, String on) {
        return anyJoin(type, "LEFT ", name, on);
    }

    @Override
    public <T extends DataObject> QueryBuilder rightJoin(Class<T> type, String name, String on) {
        return anyJoin(type, "RIGHT ", name, on);
    }

    @Override
    public QueryBuilder where(String filter, Object... args) {
        filter = processStringWithParameters(filter, args);
        
        if (this.where.length() > 0) {
            this.where.append(" AND ");
        }
        this.where.append(filter);
        
        return this;
    }

    @Override
    public QueryBuilder whereIds(Collection<Long> ids) {
        if (ids.isEmpty()) {
            throw new IllegalArgumentException();
        }
        List<String> idsText = new ArrayList<>();
        ids.stream().map(id -> id.toString()).forEach(idsText::add);
        if (this.where.length() > 0) {
            this.where.append(" AND ");
        }
        this.where.append(String.format("%s IN (%s)", this.dialect.decorateName(Constants.Data.ID_PROPERTY_NAME), String.join(",", idsText)));
        return this;
    }

    @Override
    public QueryBuilder orderBy(String column) {
        if (this.orderBy.length() > 0) {
        this.orderBy.append(String.format(", %s", this.dialect.decorateName(column)));
        } else {
            this.orderBy.append(String.format("ORDER BY %s", this.dialect.decorateName(column)));
        }
        return this;
    }

    @Override
    public QueryBuilder orderByDescending(String column) {
        if (this.orderBy.length() > 0) {
        this.orderBy.append(String.format(", %s DESC", this.dialect.decorateName(column)));
        } else {
            this.orderBy.append(String.format("ORDER BY %s DESC", this.dialect.decorateName(column)));
        }
        return this;
    }

    @Override
    public QueryBuilder groupBy(String column) {
        this.commandText.append(String.format(" GROUP BY %s ", this.dialect.decorateName(column)));
        return this;
    }

    @Override
    public QueryBuilder limit(int from, int count) {
        this.limit = this.dialect.paging(from, count);
        return this;
    }

    @Override
    public <T extends DataObject> T find(Class<T> type, long id) {
        if (commandText.length() == 0) {
            from(type);
        }
        
        return where(this.dialect.decorateName(Constants.Data.ID_PROPERTY_NAME) + " = $", id).first(type);
    }

    @Override
    public <T extends DataObject> T findOrDefault(Class<T> type, long id) {
        if (commandText.length() == 0) {
            from(type);
        }
        
        return where(this.dialect.decorateName(Constants.Data.ID_PROPERTY_NAME) + " = $", id).firstOrDefault(type);
    }
    
    private String getSql() {
        StringBuilder ret = new StringBuilder();
        ret.append("SELECT ").append(this.select).append(" ").append(this.commandText);
        
        if (this.where.length() > 0) {
            ret.append(" WHERE ").append(this.where);
        }
        
        if (this.orderBy.length() > 0) {
            ret.append(" ").append(this.orderBy);
        }
        
        if (this.limit.length() > 0) {
            ret.append(" ").append(this.limit);
        }
        
        return ret.toString();
    }
    
    private PreparedStatement prepareStatement(String sql) {
        try {
            PreparedStatement localStatement = this.connection.prepareStatement(sql);
            int parameterIndex = 1;
            for (Object param : this.parameters) {
                localStatement.setObject(parameterIndex++, param);
            }
            return localStatement;
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
    }
    
    private void readDataValue(PropertyDescriptor property, Object dataObject, ResultSet resultSet, int index) throws SQLException {
        if (property.getType().equals(OptionalInt.class)) {
            OptionalInt intValue = OptionalInt.of(resultSet.getInt(index));
            if (resultSet.wasNull()) {
                intValue = OptionalInt.empty();
            }
            property.setValue(dataObject, intValue);
        } else if (property.getType().equals(OptionalLong.class)) {
            OptionalLong longValue = OptionalLong.of(resultSet.getLong(index));
            if (resultSet.wasNull()) {
                longValue = OptionalLong.empty();
            }
            property.setValue(dataObject, longValue);
        } else if (property.getType().equals(OptionalDouble.class)) {
            OptionalDouble doubleValue = OptionalDouble.of(resultSet.getDouble(index));
            if (resultSet.wasNull()) {
                doubleValue = OptionalDouble.empty();
            }
            property.setValue(dataObject, doubleValue);
        } else if(property.getType().equals(OptionalBoolean.class)) {
            OptionalBoolean booleanValue = OptionalBoolean.of(resultSet.getBoolean(index));
            if (resultSet.wasNull()) {
                booleanValue = OptionalBoolean.empty();
            }
            property.setValue(dataObject, booleanValue);
        } else if (property.getType().equals(OffsetDateTime.class)) {
            Timestamp timestamp = resultSet.getTimestamp(index);
            if (resultSet.wasNull()) {
                property.setValue(dataObject, null);
            } else {
                OffsetDateTime dateTime = OffsetDateTime.ofInstant(timestamp.toInstant(), ZoneOffset.UTC);
                property.setValue(dataObject, dateTime);
            }
        } else {
            Object value = resultSet.getObject(index);
            if (resultSet.wasNull()) {
                property.setValue(dataObject, null);
            } else {
                property.setValue(dataObject, value);
            }
        }
    }

    private <T extends DataObject> List<T> executeOnDataObjectWithLimit(Class<T> type, String sql, OptionalInt limit, Object ... args) {
        this.parameters.clear();
        sql = processStringWithParameters(sql, args);
        List<T> ret = new ArrayList<>();
        TableMetadata metadata = DomainModelImpl.getMetadata(type);
        try (PreparedStatement statement = prepareStatement(sql);
                ResultSet resultSet = statement.executeQuery()) {
            
            ResultSetMetaData resultSetMetadata = resultSet.getMetaData();
            while (resultSet.next()) {
                
                T dataObject;
                try {
                    dataObject = type.newInstance();
                } catch (InstantiationException | IllegalAccessException ex) {
                    throw new CommonException(ex);
                }
                
                for (int i = 1; i <= resultSetMetadata.getColumnCount(); i++) {
                    String columnName = resultSetMetadata.getColumnName(i);
                    PropertyDescriptor column = metadata.getColumns().get(columnName);
                    if (column == null) {
                        continue;
                    }
                    readDataValue(column, dataObject, resultSet, i);
                }

                ret.add(dataObject);

                if (limit.isPresent() && ret.size() == limit.getAsInt()) {
                    break;
                }
            }
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
        return ret;
    }
    
    private <T> List<T> executeOnRegularObjectWithLimit(Class<T> type, String sql, OptionalInt limit, Object ... args) {
        sql = processStringWithParameters(sql, args);
        List<T> ret = new ArrayList<>();
        
        boolean isDataObject = DataObject.class.isAssignableFrom(type);
        
        Map<String, PropertyDescriptor> properties = new HashMap<>();
        
        for (PropertyDescriptor property : TypeHelper.getProperties(type)) {
            properties.put(property.getName(), property);
        }

        try (
                PreparedStatement statement = prepareStatement(sql);
                ResultSet resultSet = statement.executeQuery()) {
            
            ResultSetMetaData resultSetMetadata = resultSet.getMetaData();
            while (resultSet.next()) {
                
                T dataObject;
                if (isDataObject) {
                    try {
                        dataObject = type.newInstance();
                    } catch (InstantiationException | IllegalAccessException ex) {
                        throw new CommonException(ex);
                    }
                } else {
                    dataObject = DependencyResolver.resolve(type);
                }
                
                for (int i = 1; i <= resultSetMetadata.getColumnCount(); i++) {
                    String columnName = resultSetMetadata.getColumnName(i);
                    PropertyDescriptor property = properties.get(StringHelper.toJsonCase(columnName));
                    if (property == null) {
                        continue;
                    }
                    readDataValue(property, dataObject, resultSet, i);
                }

                ret.add(dataObject);

                if (limit.isPresent() && limit.getAsInt() == ret.size()) {
                    break;
                }
            }
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
        return ret;
    }
    
    @Override
    public <T extends DataObject> List<T> execute(Class<T> type, String sql, Object... args) {
        return executeOnDataObjectWithLimit(type, sql, OptionalInt.empty(), args);
    }

    @Override
    public boolean executeNonQuery(String sql, Object... args) {
        try (PreparedStatement localStatement = prepareStatement(sql)) {
            return localStatement.execute();
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
    }

    @Override
    public <T> T executeScalar(Class<T> type, String sql, Object... args) {
        try (PreparedStatement localStatement = prepareStatement(sql)) {
            try (ResultSet resultSet = localStatement.executeQuery()) {
                if (!resultSet.next()) {
                    return null;
                }
                return resultSet.getObject(1, type);
            }
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
    }

    @Override
    public <T> List<T> toList(Class<T> type, String selectText) {
        
        if (this.select.length() == 0) {
            this.select = selectText;
        }
        
        return executeOnRegularObjectWithLimit(type, getSql(), OptionalInt.empty());
    }

    @Override
    public List<Long> toIdList() {
        List<Long> ret = new ArrayList<>();
        try (PreparedStatement statement = prepareStatement(getSql())) {
            try (ResultSet resultSet = statement.executeQuery()) {
                while (resultSet.next()) {
                    ret.add(resultSet.getLong(Constants.Data.ID_PROPERTY_NAME));
                }
            }
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
        return ret;
    }

    @Override
    public int count() {
        this.select = "COUNT(*)";
        return (int)(long) executeScalar(Integer.class, getSql());
    }

    @Override
    public <T> T first(Class<T> type, String selectText) {
        if (selectText != null && selectText.length() > 0) {
            this.select = selectText;
        }
        
        List<T> ret = executeOnRegularObjectWithLimit(type, getSql(), OptionalInt.of(1));
        
        if (ret.isEmpty()) {
            throw new ArrayIndexOutOfBoundsException();
        }
        
        return ret.get(0);
    }

    @Override
    public <T> T firstOrDefault(Class<T> type, String selectText) {
        if (selectText != null && selectText.length() > 0) {
            this.select = selectText;
        }
        
        List<T> ret = executeOnRegularObjectWithLimit(type, getSql(), OptionalInt.of(1));
        
        if (ret.isEmpty()) {
            return null;
        }
        
        return ret.get(0);
    }

    @Override
    public QueryBuilder createNew() {
        QueryBuilderImpl ret = new QueryBuilderImpl();
        ret.connection = this.connection;
        return ret;
    }

    @Override
    public <T extends DataObject> boolean deleteFrom(Class<T> type, String filter, Object... args) {
        filter = processStringWithParameters(filter, args);
        TableMetadata metadata = DomainModelImpl.getMetadata(type);
        if (metadata == null) {
            throw new IllegalArgumentException();
        }
        String sql = String.format("DELETE FROM %s WHERE %s", this.dialect.decorateName(metadata.getName()), filter);
        try (PreparedStatement statement = prepareStatement(sql)) {
            return statement.execute();
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
    }
}

File: src/main/scala/com/lessmarkup/engine/data/QueryBuilderImpl.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.engine.data

import java.sql.{Connection, PreparedStatement, ResultSet, ResultSetMetaData, Timestamp}
import java.time.{OffsetDateTime, ZoneOffset}

import com.lessmarkup.Constants
import com.lessmarkup.engine.data.dialects.{DatabaseLanguageDialectFactory, DatabaseLanguageDialect}
import com.lessmarkup.framework.helpers._
import com.lessmarkup.interfaces.data._
import scala.collection.mutable


class QueryBuilderImpl(connection: Connection,
                       commandText: Option[String] = None,
                       select: Option[String] = None,
                       where: Option[String] = None,
                       orderBy: Option[String] = None,
                       limit: Option[String] = None,
                       parameters: List[Any] = List()) extends QueryBuilder {

  private val dialect: DatabaseLanguageDialect = DatabaseLanguageDialectFactory.createDialect(Option(this.connection))

  def join[T <: DataObject](dataType: Class[T], name: String, on: String) = anyJoin(dataType, "", name, on)

  def leftJoin[T <: DataObject](dataType: Class[T], name: String, on: String) = anyJoin(dataType, "LEFT ", name, on)

  private def anyJoin[T <: DataObject](dataType: Class[T], joinType: String, name: String, on: String): QueryBuilder = {

    val metadata: TableMetadata = MetadataStorage.getMetadata(dataType).get
    if (metadata == null) {
      throw new IllegalArgumentException
    }

    constructCopy(commandText = Option(s" $joinType ${dialect.decorateName(metadata.getName)} $name ON $on"))
  }

  def rightJoin[T <: DataObject](dataType: Class[T], name: String, on: String) = anyJoin(dataType, "RIGHT ", name, on)

  def whereIds(ids: Seq[Long]): QueryBuilder = {
    if (ids.isEmpty) {
      throw new IllegalArgumentException
    }

    constructCopy(where = Option(s"${dialect.decorateName(Constants.DataIdPropertyName)} IN (${ids.map(_.toString).mkString(",")})"))
  }

  def whereId(id: Long) = where("$+" + Constants.DataIdPropertyName + " = $", id)

  def orderBy(column: String): QueryBuilder =
    constructCopy(orderBy = Option(dialect.decorateName(column)))

  def orderByDescending(column: String): QueryBuilder =
    constructCopy(orderBy = Option(dialect.decorateName(column) + " DESC"))

  def groupBy(column: String): QueryBuilder =
    constructCopy(commandText = Option(s" GROUP BY ${dialect.decorateName(column)} "))

  def limit(from: Int, count: Int): QueryBuilder =
    constructCopy(limit = Option(dialect.paging(from, count)))

  def find[T <: DataObject](dataType: Class[T], id: Long): Option[T] =
    (if (commandText.isEmpty) from(dataType) else this)
      .where(dialect.decorateName(Constants.DataIdPropertyName) + " = $", id).first(dataType, None)

  def from[T <: DataObject](dataType: Class[T], name: Option[String]) = {

    val metadata: TableMetadata = MetadataStorage.getMetadata(dataType).get
    if (metadata == null) {
      throw new IllegalArgumentException
    }

    val safeName = name.getOrElse("")

    constructCopy(commandText = Option(s" FROM ${dialect.decorateName(metadata.getName)} $safeName"))
  }

  private def constructCopy(
                             commandText: Option[String] = None,
                             select: Option[String] = None,
                             where: Option[String] = None,
                             orderBy: Option[String] = None,
                             limit: Option[String] = None,
                             parameters: Option[List[Any]] = None) = {

    new QueryBuilderImpl(connection = this.connection,
      commandText = if (commandText.isDefined) Option(this.commandText.getOrElse("") + commandText.get) else this.commandText,
      select = if (select.isDefined) select else this.select,
      where =
        if (where.isDefined)
          if (this.where.isDefined)
            Option(this.where.get + " AND " + where.get)
          else
            where
        else
          this.where,
      orderBy =
        if (orderBy.isDefined)
          if (this.orderBy.isDefined)
            Option(this.orderBy.get + ", " + orderBy.get)
          else
            Option("ORDER BY " + orderBy.get)
        else
          this.orderBy,
      limit =
        if (limit.isDefined)
          limit
        else
          this.limit,
      parameters =
        if (parameters.isDefined)
          this.parameters ::: parameters.get
        else
          this.parameters
    )
  }

  def where(filter: String, args: Any*): QueryBuilder = {

    val ret = processStringWithParameters(filter, args)
    constructCopy(where = Option(ret._2), parameters = Option(ret._1))
  }

  private def processStringWithParameters(sql: String, args: Seq[Any]): (List[Any], String) = {
    if (args.isEmpty) {
      return (List(), sql)
    }

    val matchNumber = "\\$(\\d+)".r
    val matchTable = "\\$\\-(\\w+)".r
    val matchField = "\\$\\+(\\w+)".r
    val matchDefaultNumber = "\\$".r

    val parameters = mutable.ListBuffer[Any]()

    val allRegex = "(\\$\\-(?:\\w+))|(\\$\\+(?:\\w+))|(\\$(?:\\d*))".r

    var defaultIndex = 0

    val resultingSql = allRegex.replaceAllIn(sql, m => {
      m match {
        case matchTable(table) =>
          dialect.decorateName(MetadataStorage.getMetadata(table).get.getName)
        case matchField(field) =>
          dialect.decorateName(StringHelper.toJsonCase(field))
        case matchNumber(num) =>
          val parameterIndex = num.toInt
          parameters += args(parameterIndex)
          "?"
        case matchDefaultNumber() =>
          parameters += args(defaultIndex)
          defaultIndex += 1
          "?"

      }
    })

    (parameters.toList, resultingSql)
  }

  def execute[T <: DataObject](dataType: Class[T], sql: String, args: Any*): List[T] = {
    executeOnDataObjectWithLimit(dataType, sql, None, args: _*)
  }

  private def executeOnDataObjectWithLimit[T <: DataObject](dataType: Class[T], sql: String, limit: Option[Int], args: Any*): List[T] = {

    val sqlAndParameters = processStringWithParameters(sql, args)
    val metadata: TableMetadata = MetadataStorage.getMetadata(dataType).get

    val statement: PreparedStatement = prepareStatement(sqlAndParameters._2, sqlAndParameters._1)
    val resultSet: ResultSet = statement.executeQuery
    try {
      val resultSetMetadata: ResultSetMetaData = resultSet.getMetaData

      val stream = Stream.continually(if (resultSet.next) Option(readDataObject(dataType, resultSetMetadata, metadata, resultSet)) else None)
        .takeWhile(_.isDefined)
        .map(_.get)

      if (limit.isDefined) {
        stream.take(limit.get).toList
      } else {
        stream.toList
      }
    } finally {
      if (statement != null) statement.close()
      if (resultSet != null) resultSet.close()
    }
  }

  private def readDataObject[T <: AnyRef](dataType: Class[T], resultSetMetadata: ResultSetMetaData, metadata: TableMetadata, resultSet: ResultSet): T = {

    val dataObject: T = dataType.newInstance

    for (columnIndex <- 1 to resultSetMetadata.getColumnCount) {
      val columnName: String = resultSetMetadata.getColumnName(columnIndex)
      val column: Option[PropertyDescriptor] = metadata.getColumns.get(columnName)

      if (column.isDefined) {
        readDataValue(column.get, dataObject, resultSet, columnIndex)
      }
    }

    dataObject
  }

  def executeNonQuery(sql: String, args: Any*): Boolean = {
    val localStatement: PreparedStatement = prepareStatement(sql, parameters)
    try {
      localStatement.execute
    } finally {
      if (localStatement != null) localStatement.close()
    }
  }

  def toList[T <: AnyRef](dataType: Class[T], selectText: Option[String]): List[T] = {
    (if (this.select.isEmpty) constructCopy(select = selectText) else this)
      .executeOnRegularObjectWithLimit(dataType, getSql, None)
  }

  def toIdList: List[Long] = {
    val statement: PreparedStatement = prepareStatement(getSql, parameters)
    try {
      val resultSet: ResultSet = statement.executeQuery
      try {

        Stream.continually(if (resultSet.next) Option(resultSet.getLong(Constants.DataIdPropertyName)) else None)
          .takeWhile(_.isDefined)
          .map(_.get)
          .toList

      } finally {
        if (resultSet != null) resultSet.close()
      }
    } finally {
      if (statement != null) statement.close()
    }
  }

  def count: Int = {
    val withSelect = constructCopy(select = Option("COUNT(*)"))
    withSelect.executeScalar(classOf[Integer], withSelect.getSql).get.asInstanceOf[Long].toInt
  }

  private def getSql: String = {

    val select = this.select.getOrElse("*")
    val commandText = this.commandText.getOrElse("")
    val where = if (this.where.isDefined) " WHERE " + this.where.get else ""
    val orderBy = this.orderBy.getOrElse("")
    val limit = this.limit.getOrElse("")

    s"SELECT $select $commandText $where $orderBy $limit"
  }

  def executeScalar[T](dataType: Class[T], sql: String, args: Any*): Option[T] = {
    val localStatement: PreparedStatement = prepareStatement(sql, parameters)
    try {
      val resultSet: ResultSet = localStatement.executeQuery
      try {
        if (!resultSet.next) {
          None
        } else {
          Option(resultSet.getObject(1, dataType))
        }
      } finally {
        if (resultSet != null) resultSet.close()
      }
    } finally {
      if (localStatement != null) localStatement.close()
    }
  }

  private def prepareStatement(sql: String, parameters: List[Any]): PreparedStatement = {

    if (Constants.IsDebug) {
      LoggingHelper.getLogger(getClass).info(s"Executing statement: $sql")
    }

    val localStatement: PreparedStatement = this.connection.prepareStatement(sql)
    for ((param, index) <- parameters.view.zipWithIndex) {
      localStatement.setObject(index + 1, param)
    }
    localStatement
  }

  def first[T <: AnyRef](dataType: Class[T], selectText: Option[String]): Option[T] = {
    val builder = if (selectText.isDefined) constructCopy(select = selectText) else this
    builder
      .executeOnRegularObjectWithLimit(dataType, builder.getSql, Option(1))
      .headOption
  }

  private def executeOnRegularObjectWithLimit[T <: AnyRef](dataType: Class[T], sql: String, limit: Option[Int], args: Any*): List[T] = {

    val sqlAndParameters = processStringWithParameters(sql, args)
    val properties = TypeHelper.getProperties(dataType).toList.map(p => (p.getName, p)).toMap

    val statement: PreparedStatement = prepareStatement(sqlAndParameters._2, parameters ::: sqlAndParameters._1)
    val resultSet: ResultSet = statement.executeQuery
    try {
      val resultSetMetadata: ResultSetMetaData = resultSet.getMetaData

      val stream = Stream.continually(if (resultSet.next) Option(readRegularObject(dataType, resultSetMetadata, properties, resultSet)) else None)
        .takeWhile(_.isDefined)
        .map(_.get)

      if (limit.isDefined) {
        stream.take(limit.get).toList
      } else {
        stream.toList
      }
    } finally {
      if (statement != null) statement.close()
      if (resultSet != null) resultSet.close()
    }
  }

  private def readRegularObject[T <: AnyRef](dataType: Class[T], resultSetMetadata: ResultSetMetaData, properties: Map[String, PropertyDescriptor], resultSet: ResultSet): T = {

    val isDataObject: Boolean = classOf[DataObject].isAssignableFrom(dataType)

    val dataObject = if (isDataObject) {
      dataType.newInstance
    } else {
      DependencyResolver(dataType)
    }

    for (columnIndex <- 1 to resultSetMetadata.getColumnCount) {
      val columnName: String = resultSetMetadata.getColumnName(columnIndex)
      val property: Option[PropertyDescriptor] = properties.get(StringHelper.toJsonCase(columnName))

      if (property.isDefined) {
        readDataValue(property.get, dataObject, resultSet, columnIndex)
      }
    }

    dataObject
  }

  private def readDataValue[T <: AnyRef](property: PropertyDescriptor, dataObject: T, resultSet: ResultSet, index: Int) {
    if (property.getType == classOf[OptionInt]) {
      val intValue = resultSet.getInt(index)
      property.setValue(dataObject, if (resultSet.wasNull()) new OptionInt() else new OptionInt(intValue))
    } else if (property.getType == classOf[OptionLong]) {
      val longValue = resultSet.getLong(index)
      property.setValue(dataObject, if (resultSet.wasNull()) new OptionLong() else new OptionLong(longValue))
    } else if (property.getType == classOf[OptionDouble]) {
      val doubleValue = resultSet.getDouble(index)
      property.setValue(dataObject, if (resultSet.wasNull()) new OptionDouble() else new OptionDouble(doubleValue))
    } else if (property.getType == classOf[OptionBool]) {
      val booleanValue = resultSet.getBoolean(index)
      property.setValue(dataObject, if (resultSet.wasNull()) new OptionBool() else new OptionBool(booleanValue))
    } else if (property.getType == classOf[OptionOffsetDateTime]) {
      val timestamp: Timestamp = resultSet.getTimestamp(index)
      property.setValue(dataObject, if (resultSet.wasNull()) OptionOffsetDateTime() else OptionOffsetDateTime(OffsetDateTime.ofInstant(timestamp.toInstant, ZoneOffset.UTC)))
    } else if (property.getType == classOf[OptionString]) {
      val string: String = resultSet.getString(index)
      property.setValue(dataObject, if (resultSet.wasNull()) OptionString() else OptionString(string))
    } else if (property.getType == classOf[OptionBinaryData]) {
      val binary = resultSet.getBytes(index)
      property.setValue(dataObject, if (resultSet.wasNull()) OptionBinaryData() else OptionBinaryData(BinaryData(binary.toSeq)))
    } else if (property.getType == classOf[BinaryData]) {
      val binary = resultSet.getBytes(index)
      property.setValue(dataObject, if (resultSet.wasNull()) null.asInstanceOf[BinaryData] else BinaryData(binary.toSeq))
    } else if (property.getType == classOf[OffsetDateTime]) {
      val timestamp: Timestamp = resultSet.getTimestamp(index)
      property.setValue(dataObject, if (resultSet.wasNull()) null.asInstanceOf[OffsetDateTime]
      else {
        OffsetDateTime.ofInstant(timestamp.toInstant, ZoneOffset.UTC)
      })
    } else {
      val value: AnyRef = resultSet.getObject(index)
      if (resultSet.wasNull) {
        property.setValue(dataObject, null)
      } else {
        property.setValue(dataObject, value)
      }
    }
  }

  def createNew: QueryBuilder = constructCopy()

  def deleteFrom[T <: DataObject](dataType: Class[T], filter: String, args: Any*): Boolean = {

    val sqlWithParameters = processStringWithParameters(filter, args)
    val metadata: TableMetadata = MetadataStorage.getMetadata(dataType).get

    if (metadata == null) {
      throw new IllegalArgumentException
    }

    val sql: String = String.format("DELETE FROM %s WHERE %s", dialect.decorateName(metadata.getName), sqlWithParameters._2)

    val statement: PreparedStatement = prepareStatement(sql, sqlWithParameters._1)
    try {
      statement.execute
    } finally {
      if (statement != null) statement.close()
    }
  }

  override def decorateName(name: String): String = dialect.decorateName(name)
}
